pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,235 @@
+        stat.fragment_size()
+    } else {
+        stat.block_size()
+    };
+
+    // use blocks_available (b_avail) since, pageserver runs as unprivileged user
+    let avail_bytes = stat.blocks_available() * blocksize;
+    let total_bytes = stat.blocks() * blocksize;
+
+    // At any reasonable filesystem size, the rounding-down error caused by this integer division is negligible.
+    let usage_pct = 100 - ((100 * (avail_bytes as u64)) / (100 * (total_bytes as u64)));
+
+    let pressure = [
+        ("min_avail_bytes", avail_bytes < task_config.min_avail_bytes),
+        ("max_usage_pct", usage_pct > task_config.max_usage_pct),
+    ];
+
+    debug!(avail_bytes, total_bytes, usage_pct, ?pressure, "disk usage");
+
+    if !pressure.iter().any(|(_, problem)| *problem) {
+        return Ok(IterationOutcome::NoPressure);
+    }
+
+    warn!(
+        avail_bytes,
+        total_bytes,
+        usage_pct,
+        ?pressure,
+        "running disk usage based eviction"
+    );
+
+    // get a snapshot of the list of tenants
+    let mut data: Vec<(TenantId, Vec<(TimelineId, LocalLayerInfo)>)> =
+        crate::tenant::mgr::list_tenants()
+            .await
+            .context("get list of tenants")?
+            .into_iter()
+            .map(|(tenant_id, _)| (tenant_id, Vec::new()))
+            .collect();
+
+    // get layer infos for all timelines for all tenants
+    for (tenant_id, layer_infos) in &mut data {
+        let tenant = match crate::tenant::mgr::get_tenant(*tenant_id, false).await {
+            Ok(tenant) => tenant,
+            // this can happen if tenant has lifecycle transition after we fetched it above
+            Err(e) => {
+                warn!("failed to get tenant {tenant_id}: {e:#}");
+                continue;
+            }
+        };
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+
+        let timelines = tenant.list_timelines();
+        for tl in timelines {
+            layer_infos.extend(
+                tl.get_local_layer_infos()
+                    .into_iter()
+                    .map(|layer_infos| (tl.timeline_id, layer_infos)),
+            );
+            if cancel.is_cancelled() {
+                return Ok(IterationOutcome::Cancelled);
+            }
+        }
+    }
+
+    // sort layers by last activity
+    for (_, layers) in &mut data {
+        layers.sort_unstable_by_key(|(_, layer_info)| {
+            // XXX dedupe with eviction_task.rs?
+            let last_activity_ts = match layer_info
+                .access_stats
+                .most_recent_access_or_residence_event()
+            {
+                Either::Left(mra) => mra.when,
+                Either::Right(re) => re.timestamp,
+            };
+            last_activity_ts
+        });
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+    }
+
+    todo!("evict until we're below the threshold");
```

Comment:
Could use a compact alike itertools merge over all of the sorted lists..? Though I guess that wouldn't be fair with the 10GB reserved per tenant..

Could add a filtering to the sort_all above to only consider the least used over 10GB, then this would be stay fair? when this does not keep us below the limit, start random evictions.

Reply:
I'll resolve this conversation as the algorithm changed at the time you did this review.